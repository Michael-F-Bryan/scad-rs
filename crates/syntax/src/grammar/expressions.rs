use crate::{
    grammar::TokenSet,
    parser::{Mark, Parser},
    SyntaxKind::*,
};

const BINARY_OPERANDS: TokenSet = TokenSet::new([
    T![+],
    T![-],
    T![*],
    T![/],
    T![%],
    T![>=],
    T![>],
    T![=],
    T![<=],
    T![<],
    T![&&],
    T![||],
]);

pub(crate) fn expr(p: &mut Parser<'_>) {
    let lookahead = p.nth(1);

    match p.current() {
        T![true] | T![false] | T![undef] | STRING | INTEGER | FLOAT | IDENT
            if BINARY_OPERANDS.contains(lookahead) =>
        {
            let m = p.start();
            p.bump(p.current());
            binary_op(p, m);
        }
        T![true] | T![false] | T![undef] | STRING | INTEGER | FLOAT | IDENT
        => {
            p.bump(p.current());
        }
        L_PAREN => {
            p.bump(T!["("]);
            expr(p);
            p.expect(T![")"]);
        }
        _ => todo!(),
    }
}

fn binary_op(p: &mut Parser<'_>, m: Mark) {
    assert!(BINARY_OPERANDS.contains(p.current()));
    p.bump(p.current());

    expr(p);
    p.complete(m, BIN_EXPR);
}

#[cfg(test)]
mod tests {
    use super::*;

    parse_tests! {
        single_number => expr("42"),
        one_plus_one => expr("1 + 1"),
        parens => expr("(42)"),
        variable_lookup => expr("x"),
    }
}
