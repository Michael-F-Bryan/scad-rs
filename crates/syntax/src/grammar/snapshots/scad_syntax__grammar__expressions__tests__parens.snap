---
source: crates/syntax/src/grammar/expressions.rs
expression: children
---
[
    Token(
        L_PAREN@0..1 "(",
    ),
    Token(
        INTEGER@1..3 "42",
    ),
    Token(
        R_PAREN@3..4 ")",
    ),
]
